File: ConductorDemo/app/src/main/java/com/bluelinelabs/conductor/demo/controllers/ArchLifecycleController.java
package com.bluelinelabs.conductor.demo.controllers;

import android.arch.lifecycle.Lifecycle.Event;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bluelinelabs.conductor.ControllerChangeHandler;
import com.bluelinelabs.conductor.ControllerChangeType;
import com.bluelinelabs.conductor.RouterTransaction;
import com.bluelinelabs.conductor.archlifecycle.LifecycleController;
import com.bluelinelabs.conductor.changehandler.HorizontalChangeHandler;
import com.bluelinelabs.conductor.demo.ActionBarProvider;
import com.bluelinelabs.conductor.demo.DemoApplication;
import com.bluelinelabs.conductor.demo.R;

//import butterknife.BindView;
//import butterknife.ButterKnife;
//import butterknife.OnClick;
//import butterknife.Unbinder;

public class ArchLifecycleController extends LifecycleController {

    private static final String TAG = "ArchLifecycleController";

//    @BindView(R.id.tv_title)
    TextView tvTitle;

//    private Unbinder unbinder;
    private boolean hasExited;

    public ArchLifecycleController() {
        LifecycleObserver lifecycleObserver = new LifecycleObserver() {
            @OnLifecycleEvent(Event.ON_CREATE)
            void onCreate() {
                Log.d(TAG, "LifecycleObserver onCreate() called");
            }

            @OnLifecycleEvent(Event.ON_START)
            void onStart() {
                Log.d(TAG, "LifecycleObserver onStart() called");
            }

            @OnLifecycleEvent(Event.ON_RESUME)
            void onResume() {
                Log.d(TAG, "LifecycleObserver onResume() called");
            }

            @OnLifecycleEvent(Event.ON_PAUSE)
            void onPause() {
                Log.d(TAG, "LifecycleObserver onPause() called");
            }

            @OnLifecycleEvent(Event.ON_STOP)
            void onStop() {
                Log.d(TAG, "LifecycleObserver onStop() called");
            }

            @OnLifecycleEvent(Event.ON_DESTROY)
            void onDestroy() {
                Log.d(TAG, "LifecycleObserver onDestroy() called");
            }
        };

        Log.i(TAG, "constructor called");

        getLifecycle().addObserver(lifecycleObserver);
    }

    @NonNull
    @Override
    protected View onCreateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        Log.i(TAG, "onCreateView() called");

        View view = inflater.inflate(R.layout.controller_lifecycle, container, false);
        view.setBackgroundColor(ContextCompat.getColor(container.getContext(), R.color.orange_300));
//        unbinder = //ButterKnife.bind(this, view);

        tvTitle = view.findViewById(R.id.tv_title);
        tvTitle.setText(getResources().getString(R.string.rxlifecycle_title, TAG));

        view.findViewById(R.id.btn_next_release_view).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onNextWithReleaseClicked();
            }
        });
        view.findViewById(R.id.btn_next_retain_view).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onNextWithRetainClicked();
            }
        });
        return view;
    }

    @Override
    protected void onAttach(@NonNull View view) {
        super.onAttach(view);

        Log.i(TAG, "onAttach() called");

        (((ActionBarProvider) getActivity()).getSupportActionBar()).setTitle("Arch Components Lifecycle Demo");
    }

    @Override
    protected void onDestroyView(@NonNull View view) {
        super.onDestroyView(view);

        Log.i(TAG, "onDestroyView() called");

//        unbinder.unbind();
//        unbinder = null;
    }

    @Override
    protected void onDetach(@NonNull View view) {
        super.onDetach(view);

        Log.i(TAG, "onDetach() called");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        Log.i(TAG, "onDestroy() called");

        if (hasExited) {
            DemoApplication.refWatcher.watch(this);
        }
    }

    @Override
    protected void onChangeEnded(@NonNull ControllerChangeHandler changeHandler, @NonNull ControllerChangeType changeType) {
        super.onChangeEnded(changeHandler, changeType);

        hasExited = !changeType.isEnter;
        if (isDestroyed()) {
            DemoApplication.refWatcher.watch(this);
        }
    }

//    @OnClick(R.id.btn_next_release_view)
    void onNextWithReleaseClicked() {
        setRetainViewMode(RetainViewMode.RELEASE_DETACH);

        getRouter().pushController(RouterTransaction.with(new TextController("Logcat should now report that the Controller's onDetach() and LifecycleObserver's onPause() methods were called, followed by the Controller's onDestroyView() and LifecycleObserver's onStop()."))
                .pushChangeHandler(new HorizontalChangeHandler())
                .popChangeHandler(new HorizontalChangeHandler()));
    }

//    @OnClick(R.id.btn_next_retain_view)
    void onNextWithRetainClicked() {
        setRetainViewMode(RetainViewMode.RETAIN_DETACH);

        getRouter().pushController(RouterTransaction.with(new TextController("Logcat should now report that the Controller's onDetach() and LifecycleObserver's onPause() methods were called."))
                .pushChangeHandler(new HorizontalChangeHandler())
                .popChangeHandler(new HorizontalChangeHandler()));
    }

}

File: ConductorDemo/app/src/main/java/com/bluelinelabs/conductor/demo/controllers/ArchLifecycleController.kt
package com.bluelinelabs.conductor.demo.controllers

import android.arch.lifecycle.Lifecycle.Event
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.bluelinelabs.conductor.Controller

import com.bluelinelabs.conductor.ControllerChangeHandler
import com.bluelinelabs.conductor.ControllerChangeType
import com.bluelinelabs.conductor.RouterTransaction
import com.bluelinelabs.conductor.archlifecycle.LifecycleController
import com.bluelinelabs.conductor.changehandler.HorizontalChangeHandler
import com.bluelinelabs.conductor.demo.ActionBarProvider
import com.bluelinelabs.conductor.demo.DemoApplication
import com.bluelinelabs.conductor.demo.R

class ArchLifecycleController : LifecycleController() {

    lateinit var tvTitle: TextView

    private var hasExited: Boolean = false

    init {
        val lifecycleObserver = object : LifecycleObserver {
            @OnLifecycleEvent(Event.ON_CREATE)
            internal fun onCreate() {
                Log.d(TAG, "LifecycleObserver onCreate() called")
            }

            @OnLifecycleEvent(Event.ON_START)
            internal fun onStart() {
                Log.d(TAG, "LifecycleObserver onStart() called")
            }

            @OnLifecycleEvent(Event.ON_RESUME)
            internal fun onResume() {
                Log.d(TAG, "LifecycleObserver onResume() called")
            }

            @OnLifecycleEvent(Event.ON_PAUSE)
            internal fun onPause() {
                Log.d(TAG, "LifecycleObserver onPause() called")
            }

            @OnLifecycleEvent(Event.ON_STOP)
            internal fun onStop() {
                Log.d(TAG, "LifecycleObserver onStop() called")
            }

            @OnLifecycleEvent(Event.ON_DESTROY)
            internal fun onDestroy() {
                Log.d(TAG, "LifecycleObserver onDestroy() called")
            }
        }

        Log.i(TAG, "constructor called")

        lifecycle.addObserver(lifecycleObserver)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        Log.i(TAG, "onCreateView() called")

        val view = inflater.inflate(R.layout.controller_lifecycle, container, false)
        view.setBackgroundColor(ContextCompat.getColor(container.context, R.color.orange_300))

        tvTitle = view.findViewById(R.id.tv_title)
        tvTitle.text = resources!!.getString(R.string.rxlifecycle_title, TAG)

        view.findViewById<View>(R.id.btn_next_release_view).setOnClickListener { onNextWithReleaseClicked() }
        view.findViewById<View>(R.id.btn_next_retain_view).setOnClickListener { onNextWithRetainClicked() }
        return view
    }

    override fun onAttach(view: View) {
        super.onAttach(view)

        Log.i(TAG, "onAttach() called")

        (activity as ActionBarProvider).supportActionBar()?.title = "Arch Components Lifecycle Demo"
    }

    override fun onDestroyView(view: View) {
        super.onDestroyView(view)

        Log.i(TAG, "onDestroyView() called")
    }

    override fun onDetach(view: View) {
        super.onDetach(view)

        Log.i(TAG, "onDetach() called")
    }

    public override fun onDestroy() {
        super.onDestroy()

        Log.i(TAG, "onDestroy() called")

        if (hasExited) {
            DemoApplication.refWatcher.watch(this)
        }
    }

    override fun onChangeEnded(changeHandler: ControllerChangeHandler, changeType: ControllerChangeType) {
        super.onChangeEnded(changeHandler, changeType)

        hasExited = !changeType.isEnter
        if (isDestroyed) {
            DemoApplication.refWatcher.watch(this)
        }
    }

    private fun onNextWithReleaseClicked() {
        retainViewMode = Controller.RetainViewMode.RELEASE_DETACH

        router.pushController(RouterTransaction.with(TextController("Logcat should now report that the Controller's onDetach() and LifecycleObserver's onPause() methods were called, followed by the Controller's onDestroyView() and LifecycleObserver's onStop()."))
                .pushChangeHandler(HorizontalChangeHandler())
                .popChangeHandler(HorizontalChangeHandler()))
    }

    private fun onNextWithRetainClicked() {
        retainViewMode = Controller.RetainViewMode.RETAIN_DETACH

        router.pushController(RouterTransaction.with(TextController("Logcat should now report that the Controller's onDetach() and LifecycleObserver's onPause() methods were called."))
                .pushChangeHandler(HorizontalChangeHandler())
                .popChangeHandler(HorizontalChangeHandler()))
    }

    companion object {

        private val TAG = "ArchLifecycleController"
    }

}
